tests/cpp_unit_tests/test_tap_position_optimizer.cpp

```diff
@@ -1025,7 +1047,9 @@ TEST_CASE("Test Tap position optimizer") {
                     state_a.tap_side = tap_side;
 
                     auto optimizer = get_optimizer(strategy);
+                    auto const cached_state = state;
                     CHECK_THROWS_AS(optimizer.optimize(state, CalculationMethod::default_method), MaxIterationReached);
+                    CHECK(twoStatesEqual(cached_state, state));
```

Comment:
with the current initial tap position being `3` it actually tests nothing.

To repro: you can do some TDD-style backtracing

* checkout this branch for the tests and `origin/main` for the tap position optimizer implementation
  ```bash
  git checkout origin/main -- power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp
```
* build + run tests
  * see that they pass
  * this is not expected, because it means the original code was already compliant with this test, which should not be the case
* To debug:
  * set breakpoints before it enters the `optimizer.optimize` and in the mock transformer update function
  * step into this subcase
  * initial `tap_pos == 3`
  * it will repeatedly update between `tap_pos == 4` and `tap_pos == 3`
  * due to the even number of updates, the final `tap_pos == 3`
  * `twoStatesEqual` will therefore return `true`
* To fix:
  * set the initial to `state_b.tap_pos = 2`
    * rebuild + run the tests
    * Now, the `origin/main` version of this fails
  * restore the current branch implementation:
    * `git reset power_grid_model_c && git checkout power_grid_model_c` to checkout the current implementation on this branch
  * rebuild + run the tests
    * now, the test passes again, meaning the new implementation really is the fix for the problem